tests/code_fellows/test_deque.py

```diff
@@ -5,33 +5,76 @@
 from code_fellows.deque import Deque
 
 
-def test_deque():
+@pytest.fixture(scope="function")
+def empty_deq():
+    return Deque()
+
+
+@pytest.fixture(scope="function")
+def populated_deq():
     test_deq = Deque()
+    test_deq.appendleft("head_next")
     test_deq.appendleft("head")
-    assert test_deq.size() == 1
-    assert test_deq.peekleft() == "head"
+    test_deq.append("tail_prev")
     test_deq.append("tail")
-    assert test_deq.size() == 2
-    assert test_deq.peek() == "tail"
-    assert test_deq.peekleft() == "head"
-    assert test_deq.popleft() == "head"
-    assert test_deq.pop() == "tail"
-    assert test_deq.size() == 0
-    assert test_deq.peekleft() == None
-    assert test_deq.peekleft() == None
-    with pytest.raises(ValueError) as error:
-        test_deq.pop()
-    assert str(error.value) == "No tail to pop"
-    with pytest.raises(ValueError) as error:
-        test_deq.popleft()
-    assert str(error.value) == "No head to pop"
-    test_deq.append("new")
-    assert test_deq.head.value == "new"
-    assert test_deq.tail.value == "new"
-    assert test_deq.pop() == "new"
-    assert test_deq.size() == 0
-    test_deq.appendleft("newLeft")
-    assert test_deq.head.value == "newLeft"
-    assert test_deq.tail.value == "newLeft"
-    assert test_deq.popleft() == "newLeft"
-    assert test_deq.size() == 0
+    return test_deq
+
+
+def test_deque_appendleft_and_popleft(empty_deq):
+    empty_deq.appendleft("head")
+    assert empty_deq.size() == 1
+    assert empty_deq.peekleft() == "head"
+    assert empty_deq.popleft() == "head"
+    assert empty_deq.size() == 0
+    with pytest.raises(ValueError) as exc_info:
+        empty_deq.popleft()
+    expected = "No head to pop"
+    assert expected in str(exc_info.value)
+    assert empty_deq.peekleft() is None
+
+
+def test_deq_append_and_pop(empty_deq):
+    empty_deq.append("tail")
+    assert empty_deq.size() == 1
+    assert empty_deq.peek() == "tail"
+    assert empty_deq.pop() == "tail"
+    assert empty_deq.size() == 0
+    with pytest.raises(ValueError) as exc_info:
+        empty_deq.pop()
+    expected = "No tail to pop"
+    assert expected in str(exc_info.value)
+    assert empty_deq.peek() is None
+    assert empty_deq.__len__() == 0
+
+
+def test_deq_pop_and_size(populated_deq):
+    assert populated_deq.size() == 4
+    assert populated_deq.pop() == "tail"
+    assert populated_deq.size() == 3
+    assert populated_deq.pop() == "tail_prev"
+    assert populated_deq.size() == 2
+    # should the deque pop all the way to the head?
```

Comment:
Pop should be able to remove every item from the deque, leaving it empty.

Reply:
Oh no...
So I need to go back to the drawing board. Essentially, I have created two separate linked lists - one is created with appendleft() and the other is created by append(). But I don't have anything linking them together.

I wrote 2 new tests to check
1. append() with popleft() and peekleft()
2. appendleft() with pop() and peek()

and, of course, they are failing. 

I am not sure how to go about this now but I will start with checking if length = 0. But there is a part of me that is thinking I might have to start from scratch on this and just get rid of head and tail altogether....

Reply:
what if I kept head and tail but made my head have a tail and my tail have a head? Then I could just point the head.tail to the tail.head? 

It would still be two linked lists but connected. It would allow it to act as a proper deque, but would that be an improper construction?

Reply:
![image](https://user-images.githubusercontent.com/28361002/219463783-7f8efd72-f965-4ed7-ae92-d261bd386fb4.png)

Reply:
Your implementation needs some work.  Regardless of whether you add to one end or the other you should be able to pop from either end to get all items off the deque.

Logically, a head does not have a tail.  They are both appendages on a larger body.  So I would say no, a head shouldn't have a tail or vice versa.  This would require all nodes to have a head and tail.  

What you are describing is a doubly linked list. Where nodes have a next and previous and the list has both head and tail.  It isn't required to use this for the internal data storage, but will work just fine.  What you might not be doing correctly is maintaining all these pointers through various operations.